### DISCUSSION TITLE
Parallelising the SSCHA minimization

### DISCUSSION THREAD

**Discussion by FlemkMeserath** [SSCHAcode/python-sscha#162] [Q&A] [answered]
Dear Developers,

I was trying to use the SSCHA to obtain the phonons for a system using a large supercell.
I noticed that the SSCHA minimization becomes quite slow in this case, and I was wondering if there is any implementation already in place to parallelize the minimization of the force constants.

Thank you in advance.

**Comment by mesonepigreco** [comment, marked as answer]
Hi, 
Yes, there is a branch of the master called fourier-gradient. This implements a new algorithm for dealing with big supercells which is much faster (in some cases even a 100-1000 speedup with very large q mesh on a small primitive cell). 
This method supports parallelization, therefore it can run in parallel with mpirun (using mpi4py) parallelizing the gradient calculation over the configurations. 

I planned to merge it in the master and release 1.4 soon, but if you want, you can try the branch version (any feedback would be very much appreciated).

All the best,
Lorenzo

**Reply by FlemkMeserath** [reply]
Thankl you. I will check it out!

**Comment by FlemkMeserath** [comment]
I tried to install that specific version of the SSCHA, But unfortunately it gives me an error.
I am surrently using the most upt to date cell constructor version and I've installed the fourier-gradient version of the SSCHA.


I am running a JupyterLab notebook (The same notebook was running fine with the previous version of the SSCHA) and upon reading the following lines:
 
```
MATDYN = PREFIX + str(POPULATION-1) + '.dyn'
    
 dyn = CC.Phonons.Phonons(MATDYN, NQIRR) 
 print(dyn.GetSupercell())               
 
ens = sscha.Ensemble.Ensemble(dyn, T, SUPERCELL) 
```

I get the following error:



```
_---> 19 ens = sscha.Ensemble.Ensemble(dyn, T, SUPERCELL) 
     20 if GENERATE_ENS == True:
     21     ens.generate(N_RANDOM)       

File ~/.conda/envs/SSCHA_fourier/lib/python3.10/site-packages/sscha/Ensemble.py:189, in Ensemble.__init__(self, dyn0, T0, supercell, **kwargs)
    186 self.ignore_small_w = False
    188 # The original dynamical matrix used to generate the ensemble
--> 189 self.dyn_0 = dyn0.Copy()
    190 self.T0 = T0
    192 # This is the weight of each configuration in the sampling.
    193 # It is updated with the update_weigths function

File ~/.conda/envs/SSCHA_fourier/lib/python3.10/site-packages/sscha/Ensemble.py:305, in Ensemble.__setattr__(self, name, value)
    302     super(Ensemble, self).__setattr__(name, value)
    304 if name == "dyn_0":
--> 305     self.w_0, self.pols_0, self.w_q_0, self.pols_q_0 = value.DiagonalizeSupercell(return_qmodes=True)
    306     self.current_dyn = value.Copy()
    307     self.current_w = self.w_0.copy()

TypeError: Phonons.DiagonalizeSupercell() got an unexpected keyword argument 'return_qmodes'_
```


Any idea of what could go wrong?
